### PR TITLE
WIP: improves SelectByTargetMeanPerformance functinality

### PR DESCRIPTION
**Pull request by SunnyxBd**

### CODE REVIEW

**Comment by Morgan-Sell** [comment]
I fetched the branch. I'll get to work ;) 

To help w/ documentation, I pasted the comments/notes from the initial task below.

### INITIAL COMMENT

Idea:

Create predictor class TargetMeanPrediction or similar name with methods fit and predict.

- Fit - learns transformation
- predict - returns the mean target value per observation

Output:

- This transformer will automatically output the mean value of the target per category if variable is categorical (we have an encoder for this).
- If a variable is numerical, it will first discretize it (we have discretizers for this, equal width and frequency, user selects) and then replace by the target mean.

The reason to create a predictor class is that then, we can use it with cross_validate and cross_val_score, in the main selector function.

Things to consider:

- if categorical variables are highly cardinal, the encoders will introduce NaN. In issue expand categorical detection of NAN after encoding #294 we expand the functionality to inform in which variables nan are being introduced to help user troubleshoot
- if a numerical variable is highly skewed, nan can be introduced. In issue, add functionality to detect introduction of NAN in transform method of discretizers #295 we expand the functionality to inform in which variables nan are being introduced to help user troubleshoot

Then we need to re-code the class SelectByTargetMeanPerformance to call our predictor, and use it with cross-validate to return the important features. The advantage of using corss_validate is not just the cross_validation, which offers a less biased score, but it also allows the use of other metrics, not just roc and r2 as what we have at the moment.


### ADDT'L GUIDANCE

It requires investigating the sklearn base classes that are suitable for this estimator. We need (if it exists) the equivalent of the `TransformerMixin` but for an estimator instead of a transformer.

You can get some info from the sklearn docs. Search for "building your own estimator" and "sklearn".

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

It seems like much of the required code for the TargetMeanPredictor class exists within the existing codebase for the SelectByTargetMeanPerformance class. What's the advantage of having the encoder, discretisers, and other functionality within the TargetMeanPredictor class?

**Comment by solegalli** [comment]
The target mean predictor should have methods fit and predict.

It should take a variable, if it is categorical, output the mean value of the target per category. If it is numerical, sort the variable in bins and then replace the bins by the mean value of the target. This would be done in the predict method. The fit method learns the bins limits and the mappings.

If we pack the above code in a predictor class, with fit predict methods, then we can use that class as part of a cross-validate function.

So in short, I would like to take the functionality out of the SelectByTargetMeanPerformance into a predictor class, and then call this new predictor from the SelectByTargetMeanPerformance using cross-validation, instead of the current data split that we do at the moment.

**Comment by Morgan-Sell** [comment]
hola @solegalli,

Thanks for the explanation! 

Should the `predict()` method generate a new dataframe or replace the inputted dataframe?

**Comment by solegalli** [comment]
Predict should return a pandas series with the predictions only, which are the mean value of the target per category or bin. And this is actually a good point, because then this predictor class would only take 1 variable as input, not multiple.
